server/api/db/sqldb/db.py

```diff
@@ -871,26 +870,28 @@ def del_artifacts(
                 # record was not found
                 continue
 
-            # we do best effort deletion
-            try:
-                if artifact_column_identifier == "id":
-                    # deleting tags, because in sqlite the relationships aren't necessarily cascading
-                    self._delete(session, ArtifactV2.Tag, obj_id=column_value)
-                    self._delete(session, ArtifactV2, id=column_value)
-                else:
-                    # it's the artifact's key
-                    # deleting tags, because in sqlite the relationships aren't necessarily cascading
-                    self._delete(
-                        session, ArtifactV2.Tag, project=project, obj_name=column_value
-                    )
-                    self._delete(session, ArtifactV2, project=project, key=column_value)
-            except Exception as exc:
-                logger.warning(f"Failed to delete artifact {key}", exc_info=str(exc))
-                failed_to_delete_keys.append(key)
+            artifact_column_identifiers.setdefault(
+                artifact_column_identifier, []
+            ).append(column_value)
+
+        failed_deletions_count = 0
+        for (
+            artifact_column_identifier,
+            column_values,
+        ) in artifact_column_identifiers.items():
+            deletions_count = self._delete_multi_objects(
```

Comment:
In such case, you delete all artifact tags first, then all artifact objects. 
What happens if you delete all of the artifact tags and then try to delete the first artifact object and it fails? Now, all other artifacts remain untagged. 

Since we intend to remove the artifact data (in the future), we must first delete one artifact and its tags and data before moving on to another artifact (since deletion can fail and we need to stop).

So I suggest deleting the artifact tags and then the artifact object without commit, then moving to another artifact and finally committing everything.

Reply:
Good point but it should be a transaction anyway so the tags and artifacts should be deleted together (I believe `session.execute` does not auto commit but I'll verify.

> Since we intend to remove the artifact data (in the future), we must first delete one artifact and its tags and data before moving on to another artifact (since deletion can fail and we need to stop).

I'm not sure I understand. You will not know if the deletion failed until you commit so I think either way you have to delete all of the data and then the artifacts. If you want to delete them one by one to be able to abort you have to commit after each artifact.

Reply:
I verified that `session.execute` does not commit. Other issues were discussed offline